ibis/expr/window.py

```diff
@@ -128,32 +117,36 @@ def equals(self, other, cache=None):
         if cache is None:
             cache = {}
 
-        if (self, other) in cache:
-            return cache[(self, other)]
-
-        if id(self) == id(other):
-            cache[(self, other)] = True
+        if self is other:
+            cache[self, other] = True
             return True
 
         if not isinstance(other, Window):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
+        try:
+            return cache[self, other]
+        except KeyError:
+            pass
+
         if (len(self._group_by) != len(other._group_by) or
                 not ops.all_equal(self._group_by, other._group_by,
                                   cache=cache)):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
         if (len(self._order_by) != len(other._order_by) or
                 not ops.all_equal(self._order_by, other._order_by,
                                   cache=cache)):
-            cache[(self, other)] = False
+            cache[self, other] = False
             return False
 
-        equal = (self.preceding == other.preceding and
-                 self.following == other.following)
-        cache[(self, other)] = equal
+        equal = (
+            ops.all_equal(self.preceding, other.preceding, cache=cache) and
+            ops.all_equal(self.following, other.following, cache=cache)
+        )
+        cache[self, other] = equal
```

Comment:
It looks like the exact cached value is returned from each branch. Could this be done via a decorator?